### ISSUE TITLE
Nashorn Deprecation

### ISSUE THREAD

**Issue by agusalex** [TopQuadrant/shacl#83]
Since JDK 11 Nashorn engine is deprecated and it is planned for it to be removed. Any plan on changing the engine?  GraalVM seems to be an open source alternative.

**Comment by HolgerKnublauch** [comment]
We are monitoring the situation but I don't yet see a need to take action. Graal will definitely be a better choice, e.g. due to its latest EcmaScript support. But OTOH it is very new and Oracle has promised to keep Nashorn in the code base for many more years. So it's not really on the top of (my) priorities. If anyone has more experience in this area then I'd certainly love to hear.

**Comment by ashleycaselli** [comment]
Is there anyone to your knowledge that is taking this forward?

**Comment by HolgerKnublauch** [comment]
Sorry I don't have time to work on this right now. There still seem to be too many hurdles to get Graal going. The story would be very different if Graal would be better integrated (or anyone else knowledgeable can get this started and tell me what I need to do).

**Comment by ashleycaselli** [comment]
First of all thanks for your reply.

I started to work on it ([https://github.com/ashleycaselli/shacl/tree/develop](https://github.com/ashleycaselli/shacl/tree/develop))
I added Graal with [Nashorn Compatibility Mode](https://github.com/graalvm/graaljs/blob/master/docs/user/NashornMigrationGuide.md#nashorn-compatibility-mode) enabled using the polyglot API. 124/125 of the DASHTestCases pass

It would be nice if anyone else is interested to help for the full migration. @agusalex are you still interested?

**Comment by ashleycaselli** [comment]
A small update on this

> First of all thanks for your reply.
> 
> I started to work on it (https://github.com/ashleycaselli/shacl/tree/develop)
> I added Graal with [Nashorn Compatibility Mode](https://github.com/graalvm/graaljs/blob/master/docs/user/NashornMigrationGuide.md#nashorn-compatibility-mode) enabled using the polyglot API. 124/125 of the DASHTestCases pass
> 
> It would be nice if anyone else is interested to help for the full migration. @agusalex are you still interested?


I managed to get 125/125 on the DASHTestCases (class `TestDASHTestCasesGraal`) with Graal. The current implementation supports both Nashorn and Graal (`JSScriptEngineFactory.createScriptEngine` now accepts an `engineName`)

**Comment by HolgerKnublauch** [comment]
This sounds great! I will hopefully find the time to study this next week. Busy with some project work right now :)

**Comment by atextor** [comment]
Hi, are there any updates on this? What would be required to get this merged?

**Comment by afs** [comment]
FYI: Jena is experimenting with adopting GraalVM.
https://issues.apache.org/jira/browse/JENA-1950
(it is made complicated by license issues)

**Comment by ashleycaselli** [comment]
> Hi, are there any updates on this?

Hi @atextor. Unfortunately there has not been any update since May due to lack of time.
I was planning to work on it during the following months.

> What would be required to get this merged?

Imho, the full support to GraalVM should be added. At the moment the engine is only running with the Nashorn Compatibility Mode enabled (being backward compatible) which obviously does not allow to exploit GraalVM at its fullest.

**Comment by HolgerKnublauch** [comment]
Yes it's a tricky question how to proceed, and I admit to be delaying this for as long as possible because I am unsure at this stage. One reason is that we have in the meantime switched to a Graal-based solution in TopBraid products, with a much grander vision than SHACL-JS, see http://datashapes.org/active/ . This is now working fine but will be limited to the non-open source TopBraid products for now. Theoretically a limited version of this work could make it back into the open source package but that would be much easier if Graal would already be "bundled" with Jena somehow as introducing this dependency only for the few SHACL-JS users feels like an overkill.

I guess part of the issue is that I don't really know how many people use SHACL-JS, or whether the motivation to replace Nashorn is for other reasons such as removing the build dependency. If anyone has input...

**Comment by HolgerKnublauch** [comment]
I have decided to drop SHACL-JS support from this API. I have moved the corresponding code into our product code base but out of the open source repo to avoid the Nashorn issues.

https://github.com/TopQuadrant/shacl/commit/5f595afbfed09a2ee46f5f013c4a3f6e4a980ae2

I am well aware that an incremental route via GraalVM would have been possible but we already have a competing approach in our products via http://datashapes.org/active/ so that I don't see value in continuing these parallel efforts.
